### PR TITLE
Keep comment when modify anything in kustomization.yaml

### LINKED ISSUE #5158: kustomize edit fix deletes trailing comments
### What happened?

running command `kustomize edit fix` changes

```yaml
---
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: demo
bases:
  - ../../base
patchesStrategicMerge:
  - ./deployment.yaml
# resources:
# - ./ingress-t2.yaml
# - ./middleware-frontend-t2.yaml
# - ./middleware-backend-t2.yaml

```

to 
```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: demo
resources:
- ../../base
patches:
- path: ./deployment.yaml
```

### What did you expect to happen?

it should **keep comments untouched**

### How can we reproduce it (as minimally and precisely as possible)?

```yaml
# kustomization.yaml
---
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: demo
bases:
  - ../../base
patchesStrategicMerge:
  - ./deployment.yaml
# resources:
# - ./ingress-t2.yaml
# - ./middleware-frontend-t2.yaml
# - ./middleware-backend-t2.yaml
```

run 

`kustomize edit fix`

### Expected output

```yaml
# kustomization.yaml
---
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: demo
resources:
- ../../base
patches:
- path: ./deployment.yaml
# resources:
# - ./ingress-t2.yaml
# - ./middleware-frontend-t2.yaml
# - ./middleware-backend-t2.yaml
```

### Actual output

```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: demo
resources:
- ../../base
patches:
- path: ./deployment.yaml
```

### Kustomize version

v5.0.1

### Operating system

MacOS

### PR DESCRIPTION
**Pull request by wingyplus**
Use `kyaml/yaml/comments.CopyComments` to copy comments from the original kustomization source file to the updated file. 

Fixes #5158

### CODE REVIEW

**Comment by wingyplus** [comment]
The PR still lacks the test but I want some feedback on it. 🙇

File: kustomize/commands/internal/kustfile/kustomizationfile.go [hunk 1]
```diff
@@ -182,11 +184,49 @@ func (mf *kustomizationFile) Write(kustomization *types.Kustomization) error {
...
+	if err == io.EOF {
+		return mf.fSys.WriteFile(mf.path, result)
+	}
+	if err != nil {
+		fmt.Printf("Parse: source %v\n", err)
+		return err
+	}
+
+	if err := comments.CopyComments(from, to); err != nil {
+		fmt.Printf("CopyComments: %v\n", err)
```
Comment (wingyplus) [inline kustomize/commands/internal/kustfile/kustomizationfile.go:220]:
```suggestion
```

File: kustomize/commands/internal/kustfile/kustomizationfile.go [hunk 2]
```diff
@@ -182,11 +184,49 @@ func (mf *kustomizationFile) Write(kustomization *types.Kustomization) error {
...
+		return err
+	}
+
+	from, err := kyaml.Parse(string(data))
+	// If source is empty. Stop copy the comments.
+	if err == io.EOF {
+		return mf.fSys.WriteFile(mf.path, result)
+	}
+	if err != nil {
+		fmt.Printf("Parse: source %v\n", err)
```
Comment (wingyplus) [inline kustomize/commands/internal/kustfile/kustomizationfile.go:215]:
```suggestion
```

File: kustomize/commands/internal/kustfile/kustomizationfile.go [hunk 3]
```diff
@@ -182,11 +184,49 @@ func (mf *kustomizationFile) Write(kustomization *types.Kustomization) error {
...
+	to, err := kyaml.Parse(string(result))
 	if err != nil {
+		fmt.Printf("Parse: result %v\n", err)
 		return err
 	}
-	return mf.fSys.WriteFile(mf.path, data)
+
+	data, err := mf.fSys.ReadFile(mf.path)
+	if err != nil {
+		fmt.Printf("ReadFile: source %v\n", err)
```
Comment (wingyplus) [inline kustomize/commands/internal/kustfile/kustomizationfile.go:205]:
```suggestion
```

File: kustomize/commands/internal/kustfile/kustomizationfile.go [hunk 4]
```diff
@@ -182,11 +184,49 @@ func (mf *kustomizationFile) Write(kustomization *types.Kustomization) error {
...
+		return err
+	}
+
+	if !mf.fSys.Exists(mf.path) {
+		return mf.fSys.WriteFile(mf.path, result)
+	}
+
+	to, err := kyaml.Parse(string(result))
 	if err != nil {
+		fmt.Printf("Parse: result %v\n", err)
```
Comment (wingyplus) [inline kustomize/commands/internal/kustfile/kustomizationfile.go:199]:
```suggestion
```

**Comment by varshaprasad96** [comment]
@wingyplus could you please rebase the changes onto master.

**Comment by wingyplus** [comment]
@varshaprasad96 Sorry for reply to you lately. The PR got rebased. 

One thing I want to discuss is about behaviour change introduced by PR. The comment still keep but blank lines got consumed by `CopyComments`. Here is test that broken:


```
--- FAIL: TestPreserveCommentsWithAdjust (0.00s)
    kustomizationfile_test.go:278: Mismatch (-expected, +actual):
          bytes.Join({
        -       "\n\n\n\n",
                "# Some comments\n# This is some comment we should preserve\n# don'",
                "t delete it\n",
        -       "  ",
                "# See which field this comment goes into\nresources:\n- ../namespa",
                "ces\n- pod.yaml\n- service.yaml",
        -       "\n",
                "\napiVersion: kustomize.config.k8s.io/v1beta1\nkind: kustomization",
        -       "\n",
                "\n# something you may want to keep\nvars:\n- fieldref:\n    fieldPat",
                "h: metadata.name\n  name: MY_SERVICE_NAME\n  objref:\n    apiVersio",
                "n: v1\n    kind: Service\n    name: my-service",
        -       "\n",
                "\n# some descriptions for the patches",
        -       "\n",
                "\npatchesStrategicMerge:\n- service.yaml\n- pod.yaml\n# generator op",
                "tions\ngeneratorOptions:\n  disableNameSuffixHash: true\n",
          }, "")
--- FAIL: TestFixPatchesFieldForExtendedPatch (0.00s)
    kustomizationfile_test.go:321: Mismatch (-expected, +actual):
          bytes.Join({
        -       "\n",
                "patches:\n- path: patch1.yaml\n  target:\n    kind: Deployment\n- pa",
                "th: patch2.yaml\n  target:\n    kind: Service\napiVersion: kustomiz",
                "e.config.k8s.io/v1beta1\nkind: Kustomization\n",
          }, "")
--- FAIL: TestCommentsWithDocumentSeperatorAtBeginning (0.00s)
    kustomizationfile_test.go:368: Mismatch (-expected, +actual):
          bytes.Join({
        -       "\n\n\n",
                "# Some comments\n# This is some comment we should preserve\n# don'",
                "t delete it\napiVersion: kustomize.config.k8s.io/v1beta1\nkind: Ku",
                "stomization",
        -       "\n",
                "\nnamespace: mynamespace\n",
          }, "")
FAIL
FAIL    sigs.k8s.io/kustomize/kustomize/v5/commands/internal/kustfile   0.347s
```

If you (and reviewer) are ok with this. I can continue on this PR to fix the test.

**Comment by annasong20** [comment]
Hi @wingyplus, thanks for the PR! 

The current fix works: re-parsing the original kustomization into an `*RNode` and copying its comments. However, given that it ignores spacing and we're really only missing trailing comments after the last field (see https://github.com/kubernetes-sigs/kustomize/issues/5158#issuecomment-1604907095), a simpler solution would be to modify the existing code that parses comments, [`parseCommentedFields`](https://github.com/kubernetes-sigs/kustomize/blob/master/kustomize/commands/internal/kustfile/kustomizationfile.go#L202). For example, just thinking off the top of my head, you could 

* add a `lastComments` field to [the `kustomizationFile` struct](https://github.com/kubernetes-sigs/kustomize/blob/master/kustomize/commands/internal/kustfile/kustomizationfile.go#L112) set in [`parseCommentedFields`](https://github.com/kubernetes-sigs/kustomize/blob/master/kustomize/commands/internal/kustfile/kustomizationfile.go#L202)
* make the field in [`commentedField`](https://github.com/kubernetes-sigs/kustomize/blob/master/kustomize/commands/internal/kustfile/kustomizationfile.go#L99) optional and append a `commentedField` containing only a comment to [`originalFields`](https://github.com/kubernetes-sigs/kustomize/blob/master/kustomize/commands/internal/kustfile/kustomizationfile.go#L115C20-L115C34)

You can also add the test demonstrating the bug before you provide the fix. That way, you can check your implementation as you progress. Your reviewer can also more easily pinpoint the issue both at a glance and if they pull down your PR to explore.

Hope this helps!

**Comment by wingyplus** [comment]
Hi, @annasong20 trailing comment at the end of file is solved. I'm curious on trailing comment after field or array element. Let's see the example:

What's a position should be when we have trailing comment after key? 

Input:
```yaml
apiVersion: ...
kind:...
resources: # Trailing comment
```

Output:
```yaml
apiVersion: ...
kind:...
# Trailing comment
resources: 
```

Or 
```yaml
apiVersion: ...
kind:...
resources: # Trailing comment

And another case is trailing comment after array element.

```yaml
apiVersion: ...
kind:...
resources: 
- deployment.yaml # Trailing comment
```

Output:
```yaml
apiVersion: ...
kind:...
resources: 
- deployment.yaml # Trailing comment
```

Or 
```yaml
apiVersion: ...
kind:...
# Trailing comment
resources: 
- deployment.yaml
```

I think leaving trailing comments as it is good. Its benefit to Kpt to enable template value replacement.

**Pull request closed without merging**